app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabase.kt

```diff
@@ -591,6 +911,259 @@ class FirebaseDatabase : Database() {
+                        chatPreview(conversationId).updateChildren(lastMessage)
+                            .addOnSuccessListener {
+                                future.complete(Unit)
+                            }.addOnFailureListener { future.completeExceptionally(it) }
+                    }.addOnFailureListener { future.completeExceptionally(it) }
+            }
+            .addOnFailureListener { future.completeExceptionally(it) }
+        return future
+    }
+
+    /**
+     * Helper function to add a text message to a conversation
+     * @param conversationId id of the conversation
+     * @message that we want to add
+     * @return a future to indicate if the message was correctly added
+     */
+    private fun addChatTextMessage(
+        conversationId: String,
+        message: Message,
+    ): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+        val data = listOf<Pair<String, Any?>>(
+            "${message.timestamp}/${FirebaseKeys.CHAT_MESSAGE_SENDER}" to message.senderId,
+            "${message.timestamp}/${FirebaseKeys.CHAT_MESSAGE_CONTENT_TEXT}" to (message.content as MessageContent.Text).text,
+        ).associate { entry -> entry }
+        chatMessages(conversationId).updateChildren(data)
+            .addOnSuccessListener { // update the chat preview
+                val lastMessage = listOf<Pair<String, Any?>>(
+                    "${FirebaseKeys.CHAT_LAST_MESSAGE}/${message.timestamp}/${FirebaseKeys.CHAT_MESSAGE_SENDER}" to message.senderId,
+                    "${FirebaseKeys.CHAT_LAST_MESSAGE}/${message.timestamp}/${FirebaseKeys.CHAT_MESSAGE_CONTENT_TEXT}" to message.content.text,
+                ).associate { entry -> entry }
+                // delete the previous last message and update it
+                chatPreview(conversationId).child(FirebaseKeys.CHAT_LAST_MESSAGE).removeValue()
+                    .addOnSuccessListener {
+                        chatPreview(conversationId).updateChildren(lastMessage)
+                            .addOnSuccessListener {
+                                future.complete(Unit)
+                            }.addOnFailureListener { future.completeExceptionally(it) }
+                    }.addOnFailureListener { future.completeExceptionally(it) }
+            }
+            .addOnFailureListener { future.completeExceptionally(it) }
+        return future
+    }
+
+    /**
+     * Helper function to transform some data into a message
+     * @param data datasnpshot that contains the data of the message
+     * @return a message correposnding to this data
+     */
+    private fun getMessageFromData(data: DataSnapshot): Message {
+        val dateStr = data.key!!
+        val date = dateStr.toLong()
+        val sender = data.child(FirebaseKeys.CHAT_MESSAGE_SENDER).value as String
+        val dataImage = data.child(FirebaseKeys.CHAT_MESSAGE_CONTENT_IMAGE)
+        val dataRun = data.child(FirebaseKeys.CHAT_MESSAGE_CONTENT_RUN)
```

Comment:
Suggestion: Its generally good to avoid using "!!" not-null checks and instead use a safet alternative such as "let".
For example:
val dataRun = data.child(FirebaseKeys.CHAT_MESSAGE_CONTENT_RUN)
val run = dataRun.children.firstOrNull()?.let { transformRun(it) }
if (run != null) {